test/test_util.js

```diff
@@ -1,213 +1,109 @@
-        focus: noop,
-      },
-    );
-    input.simulate('change', changeEvent);
-  }
-}
+  if (specialKeys.includes(key)) {
+    if (start === end) {
+      await input.focus();
+      await input.setSelectionRange(start, end);
 
-export async function render(elm) {
-  const view = testRender(elm);
-  const input = await view.getByRole('textbox');
-  return { ...view, view: view, input };
-}
+      await user.keyboard(key);
+    } else {
+      let newValue;
 
-export function simulateNativeKeyInput(input, key, selectionStart = 0, selectionEnd = 0) {
-  input.setSelectionRange(selectionStart, selectionEnd);
-  userEvent.type(input, key);
-}
+      if (key === '{Backspace}') {
+        newValue = v.slice(0, start) + v.slice(end, v.length);
+      } else if (key === '{Delete}') {
+        newValue = v.slice(0, start) + v.slice(end, v.length);
+      }
 
-export function simulatePaste(input, test, selectionStart = 0, selectionEnd = 0) {
-  input.setSelectionRange(selectionStart, selectionEnd);
-  userEvent.paste(input, test);
-}
+      fireEvent.change(input, { target: { value: newValue } });
+    }
+  } else {
+    if (start === end) {
+      if (options?.eventType === 'keyboard') {
+        input.focus();
+        input.selectionStart = start;
+        input.selectionEnd = end;
+
+        await user.keyboard(key);
+      } else {
+        await user.type(input, key, { initialSelectionStart: start, initialSelectionEnd: end });
+      }
+    } else {
+      let newValue;
 
-export function simulateNativeMouseUpEvent(input, selectionStart) {
-  input.setSelectionRange(selectionStart, selectionStart);
-  userEvent.click(input);
+      newValue = v.slice(0, start) + key + v.slice(end, v.length);
+      fireEvent.change(input, { target: { value: newValue } });
+      end = start;
+    }
+  }
 }
 
-export function simulateMousUpEvent(input, selectionStart, setSelectionRange) {
+export function simulateMouseUpEvent(user, input, selectionStart) {
   const selectionEnd = selectionStart;
 
-  const currentValue = input.prop('value');
-
-  const mouseUpEvent = getEvent(
-    {},
-    {
-      value: currentValue,
-      selectionStart,
-      selectionEnd,
-      setSelectionRange: setSelectionRange || noop,
-      focus: noop,
-    },
-  );
-
-  input.simulate('mouseup', mouseUpEvent);
+  fireEvent.mouseUp(input, {
+    target: { selectionStart, selectionEnd },
+  });
 }
 
-export function simulateFocusEvent(input, selectionStart = 0, selectionEnd, setSelectionRange) {
-  if (selectionEnd === undefined) {
-    selectionEnd = selectionStart;
-  }
-
-  const currentValue = input.prop('value');
-
-  const focusEvent = getEvent(
-    {
-      persist: persist.bind(null, 'focus'),
-    },
-    {
-      value: currentValue,
-      selectionStart,
-      selectionEnd,
-      setSelectionRange: setSelectionRange || noop,
-      focus: noop,
-    },
-  );
-
-  input.simulate('focus', focusEvent);
+export function simulateFocusEvent(input) {
+  input.focus();
```

Comment:
**Suggestion/Test Bug**
So there are three cases how focus can happen.
- Due to mouse click on specific caret position. In which case the caret is placed at specific position. Which I think the usage is replaced now with simulateMousUpEvent.
- Due to tab press, in which case whole input value gets selected.
- Manual input.focus call, in which case also whole input value gets selected.

If we catering all the three cases with this util, we should have selectionStart and selectionEnd as param.

In case we are catering 2nd and 3rd, we need to manually set the selectionStart to 0 and selectionEnd at input.value.length

Reason: input.focus() in JSDom puts the caret position at the end, which is not same as browser.